src/main/java/net/minestom/server/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,366 @@
+package net.minestom.server.inventory;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.minestom.server.data.Data;
+import net.minestom.server.data.DataContainer;
+import net.minestom.server.inventory.click.InventoryClickProcessor;
+import net.minestom.server.inventory.condition.InventoryCondition;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.item.StackingRule;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.UnaryOperator;
+
+/**
+ * Represents an inventory where items can be modified/retrieved.
+ */
+public abstract class AbstractInventory implements InventoryClickHandler, DataContainer {
+
+    private final int size;
+    protected final ItemStack[] itemStacks;
+
+    // list of conditions/callbacks assigned to this inventory
+    protected final List<InventoryCondition> inventoryConditions = new CopyOnWriteArrayList<>();
+    // the click processor which process all the clicks in the inventory
+    protected final InventoryClickProcessor clickProcessor = new InventoryClickProcessor();
+
+    private Data data;
+
+    protected AbstractInventory(int size) {
+        this.size = size;
+        this.itemStacks = new ItemStack[getSize()];
+
+        Arrays.fill(itemStacks, ItemStack.AIR);
+    }
+
+    /**
+     * Sets an {@link ItemStack} at the specified slot and send relevant update to the viewer(s).
+     *
+     * @param slot      the slot to set the item
+     * @param itemStack the item to set
+     */
+    public abstract void setItemStack(int slot, @NotNull ItemStack itemStack);
+
+    protected abstract void safeItemInsert(int slot, @NotNull ItemStack itemStack);
+
+    /**
+     * Adds an {@link ItemStack} to the inventory and sends relevant update to the viewer(s).
+     * <p>
+     * Even the item cannot be fully added, the amount of {@code itemStack} will be updated.
+     *
+     * @param itemStack the item to add
+     * @return true if the item has been successfully fully added, false otherwise
+     */
+    public synchronized boolean addItemStack(@NotNull ItemStack itemStack) {
```

Comment:
If `ItemStack`s are immutable, shouldn't this return an `ItemStack`, perhaps being nullable to indicate it fit completely?

Reply:
What should the `ItemStack` represent? for example the item might merge with multiple items inside the inventory. We could replace it with a list of affected item, but I am unsure about its usefulness.

The method has been modified to "fail" when the item cannot completely fit, as it is error prone

Reply:
The Javadoc 1:states 
> Even the item cannot be fully added, the amount of {@code itemStack} will be updated.

Is this wrong then?

Reply:
Oops, fixed